### PR TITLE
Reorder UI, add panner, add hide button

### PR DESCRIPTION
**Pull request by alshakero**
I added the SVG for the hiding button, if you want to use the same button as a show/hide toggle, you can just hide/show
`<line stroke-width="50" x1="300" y1="0" x2="60" y2="350" />
` 
element inside the SVG:

```
<svg x="0px" y="0px" width="16px" height="16px" viewBox="0 -20 300 400">
    <path d="M359 204c4 -2 6 -6 6 -11c0 -4 -2 -8 -4 -10v0v0l-1 -1l-79 -79v0c-24 -27 -60 -44 -99 -44c-35 0 -66 14 -90 36v-1l-87 87c-3 3 -5 7 -5 11c0 3 1 7 3 9v0l80 79c24 27 60 45 99 45c35 0 67 -14 91 -36v1zM182 102c50 0 90 40 90 90s-40 90 -90 90s-90 -40 -90 -90 s40 -90 90 -90zM139 192c0 29 14 43 43 43s44 -14 44 -43s-15 -44 -44 -44s-43 15 -43 44z"></path>
   --> <line stroke-width="50" x1="300" y1="0" x2="60" y2="350" />
</svg>
```

### CODE REVIEW

File: starcounter-layout-html-editor.html [hunk 1]
```diff
@@ -219,6 +240,45 @@
...
+                    const rightPanel = this.$.rightPanel;
+                    
+                    panner.addEventListener('mousedown', function(ev) {
+                        that.isMouseDown = true;
+                    });
+
+                    /* reset on double click */
+                    panner.addEventListener('dblclick', function(ev) {
+                        leftPanel.style.flexGrow = 0.5;
+                        rightPanel.style.flexGrow = 0.5;
```
Comment (tomalec) [inline starcounter-layout-html-editor.html]: Why did you add event listener to the entire window?

File: juicy-simple-visual-html-editor.html [hunk 2]
```diff
@@ -13,16 +13,18 @@
 <dom-module id="juicy-simple-visual-html-editor">
     <template>
         <style>
-            :host{
-                min-height: 1em;
-                display: flex;
-                flex-direction: column;
+           juicy-dom-tree-view{
```
Comment (tomalec) [inline juicy-simple-visual-html-editor.html]: Why have you removed `:host` styles? At least, we need it to be non inline element by default. min height also makes sense to me

**Comment by tomalec** [comment]
Please attach listeners Polymer way https://meowni.ca/posts/polymer-cheatsheet/#listeners

File: starcounter-layout-html-editor.html (hunk 1 above)
Comment (alshakero) [inline starcounter-layout-html-editor.html]: Because when you move the mouse to drag, you can sometimes miss the _panner_ div, and it will not be the event target, and your callback wouldn't be called. You need the dragging to work even if your pointer leaves the div, tor example when you go below/above it.

File: juicy-simple-visual-html-editor.html (hunk 2 above)
Comment (alshakero) [inline juicy-simple-visual-html-editor.html]: Because when you set `display` to `flex` the tools will not be fixed and this will happen 
image
and if you scroll even more you won't be able to use the tools for lower nodes. And I couldn't see the need of minimum height, but wouldn't be of harm of course. 

That's because when display is flex you can't say overflow to hidden (http://stackoverflow.com/questions/21515042/scrolling-a-flexbox-with-overflowing-content)

Comment (alshakero) [inline juicy-simple-visual-html-editor.html]: I'd be happy if you suggest something.

File: juicy-simple-visual-html-editor.html [hunk 3]
```diff
@@ -14,15 +14,20 @@
     <template>
         <style>
             :host{
-                min-height: 1em;
-                display: flex;
-                flex-direction: column;
+                min-height: 2em;
```
Comment (tomalec) [inline juicy-simple-visual-html-editor.html]: I still don't get the point why flex is no good in here
image
We do not need `overflow:hidden` here. I'd say `display:inline` is even worse.

File: starcounter-layout-html-editor.html [hunk 4]
```diff
@@ -220,6 +271,9 @@
                         window.dispatchEvent(new Event('resize')); //fix resizing of ACE edior
                     });
                     this.refreshSelectors();
+
+                    window.addEventListener('mouseup', this.windowMouseUp.bind(this));
```
Comment (tomalec) [inline starcounter-layout-html-editor.html]: If you are attaching listeners manually you need to take care of detaching as well

File: starcounter-layout-html-editor.html [hunk 5]
```diff
@@ -98,7 +119,37 @@
...
                 is: 'starcounter-layout-html-editor',
+                listeners: {
+                    'panner.mousedown': 'pannerOnMouseDown',
+                    'panner.dblclick': 'pannerDoubleClick'
+                },
+                pannerOnMouseDown: function() {
+                    this.isMouseDown = true;
+                },
+                windowMouseMove: function(ev) {
+                    if(this.isMouseDown)
```
Comment (tomalec) [inline starcounter-layout-html-editor.html]: Have you tried drag events?

**Comment by tomalec** [comment]
In general I'd prefer to use `em` rather than `px` as then it could behave nicer if you would style entire context

File: starcounter-layout-html-editor.html (hunk 1 above)
Comment (tomalec) [inline starcounter-layout-html-editor.html]: `drag` events should refer correct target even when moved ouside

File: juicy-simple-visual-html-editor.html (hunk 3 above)
Comment (alshakero) [inline juicy-simple-visual-html-editor.html]: Can you scroll down the tree and share a screenshot?

File: starcounter-layout-html-editor.html (hunk 5 above)
Comment (alshakero) [inline starcounter-layout-html-editor.html]: Trying now. They were always pain in the neck, I'm giving them another shot if you think that's better.

File: juicy-simple-visual-html-editor.html (hunk 3 above)
Comment (tomalec) [inline juicy-simple-visual-html-editor.html]: image

**Comment by tomalec** [comment]
I'd also add some white-space between separator and the editors, in general I don't find it intuitive that it's draggable, maybe we could get some inspiration from  https://material.io/guidelines/layout/split-screen.html

Could you make those tile collapsible by separate button/arrow like https://material.io/guidelines/components/expansion-panels.html
Personally I would find it nicer to collapse it by single click, rather than dragging it to the very end.
Also once collapsed, we could hide the divider, and just leave small icon & vertical text to expand

**Comment by alshakero** [comment]
@tomalec you're right, it's not intuitive, I'll look at the reference you mentioned. 

As for collapsing the panels. I actually implemented it at first. But I figured it's a bit more complicated than its worth. Think about how many possibilities there needs to be.

1. Default state, both visible, the separator is in the middle, show both collapse buttons in _hide_ mode. Easy.
2. One is collapsed, the other should hide its collapse button, to avoid collapsing both. 
3. Both are visible, but not 50/50%. Say 30/70%. Then one collapse is clicked. We should make the visible one 100% and preserve the 30/70% when the user shows the hidden one again. 

And at least 3 other scenarios for a functionality that is already there by simply dragging. 

If you really like a collapse option, I think a separate radio group would be perfect. Something like:

- [x]  Show both
- [ ]  Show right
- [ ]  Show left




Way easier to implement, what do you think?

**Comment by tomalec** [comment]
Sorry, I don't get what you mean by all 1-3, maybe HO call could help.

I still don't get why it's complicated, but for sure I find yet another radio button very unintuitive.

**Comment by warpech** [comment]
What is the status of this? Should it wait until @tomalec is back from vacation?

**Comment by alshakero** [comment]
@warpech oh, it was a rush and both @tomalec and I implemented the collapsing feature by accident. However, I also implemented the drag resize with it, but it was on Wednesday and Tomek preferred to use his code rather than reviewing mine for the tight schedule and his vacation approaching.

**Comment by warpech** [comment]
So I propose that we try to finish PR #13 and release a version of CompositionEditor with it to meet the schedules.

Then you can decide about PR #12 when @tomalec is back from vacation. 

WDYT?

**Comment by alshakero** [comment]
Closing as implimented by @tomalec .
